sandbox/lua/decoders/json.lua

```diff
@@ -0,0 +1,116 @@
+-- This Source Code Form is subject to the terms of the Mozilla Public
+-- License, v. 2.0. If a copy of the MPL was not distributed with this
+-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+--[[
+Parses a payload containing JSON.
+
+Config:
+
+- type (string, optional, default "json"):
+    Sets the message 'Type' header to the specified value.
+
+- payload_keep (bool, optional, default false)
+    Always preserve the original log line in the message payload.
+
+.. code-block:: javascript
+
+    {
+      "msg": "Start Request",
+      "event": "artemis.web.ensure-running",
+      "extra": {
+        "workspace-id": "cN907xLngi"
+      },
+      "time": "2015-05-06T20:40:05.509926234Z",
+      "severity": 1
+    }
+
+*Example Heka Configuration*
+
+.. code-block:: ini
+
+    [ArtemisLogInput]
+    type = "LogstreamerInput"
+    log_directory = "/srv/artemis/current/logs"
+    file_match = 'artemis\.log'
+    decoder = "JsonDecoder"
+
+    [JsonDecoder]
+    type = "SandboxDecoder"
+    filename = "lua_decoders/json.lua"
+
+        [JsonDecoder.config]
+        type = "artemis"
+        payload_keep = true
+
+*Example Heka Message*
+
+:Timestamp: 2015-05-06 20:40:05 -0800 PST
+:Type: artemis
+:Hostname: test.example.com
+:Pid: 0
+:UUID: 8e414f01-9d7f-4a48-a5e1-ae92e5954df5
+:Payload:
+:EnvVersion:
+:Severity: 1
+:Fields:
+    | name:"msg" value_type:STRING value_string:"Start Request"
+    | name:"event" value_type:STRING value_string:"artemis.web.ensure-running"
+    | name:"extra.workspace-id" value_type:STRING value_string:"cN907xLngi"
+    | name:"time" value_type:STRING value_string:"2015-05-06T20:40:05.509926234Z"
+    | name:"severity" value_type:INTEGER value_integer:1
+--]]
+
+require "cjson"
+local util = require("util")
+
+local msg_type     = read_config("type") or "json"
+local payload_keep = read_config("payload_keep")
+
+local msg = {
+    Type       = msg_type,
+    Payload    = nil,
+    Fields     = nil,
+    Severity   = nil
+}
+
+function process_message()
+    local ok, json = pcall(cjson.decode, read_message("Payload"))
+    if not ok then
+        return -1
+    end
+
+    -- keep payload, or not
+    if payload_keep then
+        msg.Payload = read_message("Payload")
+    end
+
+    -- map to heka native message fields
+    local field_type_map = {
+        string = { "Uuid", "Type", "Logger", "Hostname" },
+        number = { "Severity", "EnvVersion", "Pid" }
+    }
+
+    for k, v in pairs(field_type_map) do
+        for i, f in ipairs(v) do
+            local low_f = f:lower()
+
+            if type(json[low_f]) == k then
+                msg[f] = json[low_f] ; json[low_f] = nil
+            end
+        end
+    end
```

Comment:
Not quite sure what's going on in here. Seems like we're lower-casing the values in the `field_type_map` table, but we hard-coded those values, we could have just made them lower case in the first place.

What we really want to do is to iterate through the top level values in the `json` table. We lower case every key, check to see if the lower-case value matches one of the lower-case message field names, with the right type. If so, that data gets written to that message field instead of as a dynamic field.

This will clearly cost some cycles, especially when the incoming JSON structure is large, so it might be nice to be able to turn it off with a config setting.